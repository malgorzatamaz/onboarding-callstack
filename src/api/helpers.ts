import { SerializedError } from "@reduxjs/toolkit";
import { FetchBaseQueryError } from "@reduxjs/toolkit/query";

export function getErrorMessage(
  error: SerializedError | FetchBaseQueryError
): string {
  const isFetchBaseQueryError =
    typeof error === "object" && error != null && "status" in error;

  const isErrorWithMessage =
    typeof error === "object" &&
    error != null &&
    "message" in error &&
    typeof error.message === "string";

  if (isFetchBaseQueryError) {
    return "error" in error ? error.error : JSON.stringify(error.data);
  } else if (isErrorWithMessage) {
    return error.message || "Something went wrong";
  }

  return "Something went wrong";
}
