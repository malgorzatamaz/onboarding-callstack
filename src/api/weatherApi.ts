import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { Weather, WeatherResponse } from "src/types";

//Should be the env variable
const API_KEY = "f551018c63bc5e28c8f313c99c0b43d5";

export const weatherApi = createApi({
  reducerPath: "weatherApi",
  baseQuery: fetchBaseQuery({
    baseUrl: `https://api.openweathermap.org`,
  }),
  endpoints: (builder) => ({
    getWeatherByCity: builder.query<Weather, string>({
      query: (name) =>
        `/data/2.5/weather?q=${name}&appid=${API_KEY}&units=metric`,
      transformResponse: (response: WeatherResponse): Weather => ({
        cloudCover: response?.clouds?.all,
        description: response?.weather?.[0]?.description,
        humidity: response?.main.humidity,
        pressure: response?.main.pressure,
        temperature: response?.main.temp,
        windSpeed: response?.wind?.speed,
      }),
    }),
  }),
});

export const { useGetWeatherByCityQuery } = weatherApi;
