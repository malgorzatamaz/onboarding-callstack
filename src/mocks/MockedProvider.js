import { getDefaultMiddleware } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import createMockStore from "redux-mock-store";
import React from "react";

const middlewares = getDefaultMiddleware();
const mockStore = createMockStore(middlewares);
const store = mockStore();

export const MockedProvider = ({ children }) => {
  return <Provider store={store}>{children}</Provider>;
};
