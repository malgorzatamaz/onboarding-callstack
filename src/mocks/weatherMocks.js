export const weatherMocks = {
  cloudCover: 75,
  description: "sunny",
  humidity: 60,
  pressure: 1015,
  temperature: 15,
  windSpeed: 3,
};
