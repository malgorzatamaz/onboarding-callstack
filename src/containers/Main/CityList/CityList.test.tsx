import { getDefaultMiddleware } from "@reduxjs/toolkit";
import { NavigationContainer } from "@react-navigation/native";
import { Provider } from "react-redux";
import { render } from "@testing-library/react-native";
import createMockStore from "redux-mock-store";
import React from "react";

const middlewares = getDefaultMiddleware();
const mockStore = createMockStore(middlewares);
const store = mockStore();

import * as hooks from "src/api/weatherApi";
import { weatherMocks } from "src/mocks/weatherMocks";
import { CityList } from "./";

jest.spyOn(hooks, "useGetWeatherByCityQuery").mockReturnValue({
  data: weatherMocks,
  isError: false,
  isLoading: false,
  refetch: jest.fn(),
});

test("should render weather details", async () => {
  const { getByText, queryAllByText, toJSON } = render(
    <Provider store={store}>
      <NavigationContainer>
        <CityList />
      </NavigationContainer>
    </Provider>
  );

  expect(toJSON()).toMatchSnapshot();
  expect(getByText("Warszawa")).toBeTruthy();
  expect(getByText("Wrocław")).toBeTruthy();
  expect(getByText("Żywiec")).toBeTruthy();
  expect(queryAllByText("15 °C")).toHaveLength(3);
});
