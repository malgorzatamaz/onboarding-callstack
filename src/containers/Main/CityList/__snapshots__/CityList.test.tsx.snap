// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render weather details 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      "Warszawa",
      "Wrocław",
      "Żywiec",
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "height": 60,
            "padding": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000",
                "flex": 1,
              }
            }
          >
            Warszawa
          </Text>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "justifyContent": "center",
              }
            }
          >
            Badge
          </View>
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/next.png",
              }
            }
            style={
              Object {
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
      </View>
      Divider
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "height": 60,
            "padding": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000",
                "flex": 1,
              }
            }
          >
            Wrocław
          </Text>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "justifyContent": "center",
              }
            }
          >
            Badge
          </View>
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/next.png",
              }
            }
            style={
              Object {
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
      </View>
      Divider
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "height": 60,
            "padding": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000",
                "flex": 1,
              }
            }
          >
            Żywiec
          </Text>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "justifyContent": "center",
              }
            }
          >
            Badge
          </View>
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/next.png",
              }
            }
            style={
              Object {
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
